util: remove kubernetes csi prefixed parameters

Madhu-1 opened the pull request:
remove kubernetes csi prefixed parameters from the volumeContext as we dont want to store it in the PV VolumeAttributes.
    
updates: #2835 

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
util: add helper function to strip kube parameters

added helper function to strip the kubernetes
specific parameters from the volumeContext as
volumeContext is storaged in the PV volumeAttributes

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: remove kubernetes csi prefixed parameters

remove kubernetes csi prefixed parameters
from the volumeContext as we dont want
to store it in the PV VolumeAttributes.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
cephfs: remove kubernetes csi prefixed parameters

remove kubernetes csi prefixed parameters
from the volumeContext as we dont want
to store it in the PV VolumeAttributes.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

pkalever commented:
@Madhu-1 I hope the below values won't be removed
```
// PV metadata keys
pvcNameKey      = "csi.storage.k8s.io/pvc/name"
pcNamespaceKey = "csi.storage.k8s.io/pvc/namespace"
pvNameKey       = "csi.storage.k8s.io/pv/name"

// snapshot metadata keys                                              
volSnapNameKey        = "csi.storage.k8s.io/volumesnapshot/name"       
volSnapNamespaceKey   = "csi.storage.k8s.io/volumesnapshot/namespace"  
volSnapContentNameKey = "csi.storage.k8s.io/volumesnapshotcontent/name"
```

https://github.com/ceph/ceph-csi/pull/2898

Madhu-1 commented:
> @Madhu-1 I hope the below values won't be removed
> 
> ```
> // PV metadata keys
> pvcNameKey      = "csi.storage.k8s.io/pvc/name"
> pcNamespaceKey = "csi.storage.k8s.io/pvc/namespace"
> pvNameKey       = "csi.storage.k8s.io/pv/name"
> 
> // snapshot metadata keys                                              
> volSnapNameKey        = "csi.storage.k8s.io/volumesnapshot/name"       
> volSnapNamespaceKey   = "csi.storage.k8s.io/volumesnapshot/namespace"  
> volSnapContentNameKey = "csi.storage.k8s.io/volumesnapshotcontent/name"
> ```
> 
> #2898

This is only for PVC not for snapshot. Yes it will be removed. All keys starting with `csi.storage.k8s.io` will be removed from the volumeAttributes.

pkalever commented:
> This is only for PVC not for snapshot. Yes it will be removed. All keys starting with `csi.storage.k8s.io` will be removed from the volumeAttributes.

This will affect the PR https://github.com/ceph/ceph-csi/pull/2898
Why do we want to do this?

Madhu-1 commented:
> > This is only for PVC not for snapshot. Yes it will be removed. All keys starting with `csi.storage.k8s.io` will be removed from the volumeAttributes.
> 
> This will affect the PR #2898 Why do we want to do this?

How it will be affected?

pkalever commented:
> How it will be affected?

As you know, the mentioned PR tries to read the details of PV and snapshot from the volume attributes, If you think this won't affect, I doubt I missing something.

Madhu-1 commented:
> > How it will be affected?
> 
> As you know, the mentioned PR tries to read the details of PV and snapshot from the volume attributes, If you think this won't affect, I doubt I missing something.

This will not affect the PR you are working on looks like it affect the encryption as PVC Namespace is required. will re-check on this one.

Madhu-1 committed:
rbd: do not read pvc namespace from volume attributes

Below are the 3 different cases where we need
the PVC namespace for encryption

* CreateVolume:- Read the namespace from the
createVolume parameters and store it in the omap
* NodeStage:- Read the namespace from the omap
not from the volumeContext
* Regenerate:- Read the pvc namespace from the claimRef
not from the volumeAttributes.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
/retest ci/centos/mini-e2e/k8s-1.23

Madhu-1 commented:
> /retest ci/centos/mini-e2e/k8s-1.23

`    [91mMar 16 09:16:31.643: failed to delete PVC and Deployment: timed out waiting for the condition[0m
`
Failed due to cephfs error not related to this PR.

Madhu-1 commented:
@nixpanic @Rakshith-R PTAL from an encryption point of view for double confirmation.

Rakshith-R commented:
There might be a problem when tenant based encryption is being used and pvc created in namespace A is recreated in namespace B.

This change will override the namespace in omap and we may not be able to unlock the volume.

Before the change, reading ns from pv spec ensured the namespace remained the same(in omap).

Madhu-1 commented:
> There might be a problem when tenant based encryption is being used and pvc created in namespace A is recreated in namespace B.
> 
> This change will override the namespace in omap and we may not be able to unlock the volume.
> 

This will not override the namespace in the omap, the owner will be retrieved from the OMAP in case of normal workflow (NodeStage). in case of DR, the omap will be regenerated with the correct owner specified in the PV claimRef object. 

> Before the change, reading ns from pv spec ensured the namespace remained the same(in omap).

IMO This is a wrong assumption, what if the whole namespace is deleted and PV is reattached to a new PVC in a different namespace. in the case of DR restoring to a new namespace should be a supported option. and more over different values in different places in the PV is also not correct.

nixpanic commented:
> @nixpanic @Rakshith-R PTAL from an encryption point of view for double confirmation.

Encryption uses `owner` (maps to the Kubernetes Namespace), which is set as metadata on the RBD image.

I agree that we should not store any `csi.storage.k8s.io` keys, not in the journal or metadata. We should use non-Kubernetes names for attributes that we need, and explicitly add those where needed.

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -248,8 +248,8 @@ func (rv *rbdVolume) openEncryptedDevice(ctx context.Context, devicePath string)
 
-func (ri *rbdImage) initKMS(ctx context.Context, volOptions, credentials map[string]string) error {
-	kmsID, err := ri.ParseEncryptionOpts(ctx, volOptions)
+func (ri *rbdImage) initKMS(ctx context.Context, namespace string, volOptions, credentials map[string]string) error {
```
better make sure to set `rbdImage.Owner` before calling this, no need to pass a `namespace` parameter that should have the same contents as the `.Owner` attribute.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -119,17 +120,24 @@ func (cs *ControllerServer) parseVolCreateRequest(
 	}
 
+	// get the owner of the PVC which is required for few encryption related operations
+	namespace := req.GetParameters()["csi.storage.k8s.io/pvc/namespace"]
```
Namespace is Kubernetes specific. Better to call this owner?
It is cleaner to set the `rbdVol.Owner` as part of `genVolFromVolumeOptions()`

You also added a `k8s/parameters.go` file, make sure to move the `csi.storage.k8s.io/pvc/namespace` string there too, and maybe create a function `k8s.GetOwner(req.GetParameters())`.

Madhu-1 replied:
>It is cleaner to set the rbdVol.Owner as part of genVolFromVolumeOptions()

This is not possible because of in node plugin we are extracting the owner details from omap after calling  `genVolFromVolumeOptions` that is why i moved the  setting owner and calling `rv.initKMS` outside of the `genVolFromVolumeOptions` function

other changes are done except above

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -235,6 +236,11 @@ func populateRbdVol(
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	err = rv.initKMS(ctx, rv.Owner, req.GetVolumeContext(), req.GetSecrets())
```
passing `rv.Owner` makes no sense here, the `rv.initKMS()` function can access `rv.Owner` itself.

Madhu-1 committed:
util: add helper to get pvcnamespace from input

added helper function to return the pvc namespace
name from the input parameters.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.22

Madhu-1 commented:
> /test ci/centos/mini-e2e-helm/k8s-1.22

```
deployment.apps/snapshot-controller created

customresourcedefinition.apiextensions.k8s.io/volumesnapshotclasses.snapshot.storage.k8s.io created

Error from server: error when creating "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v5.0.1/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml[](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/2312/pipeline#step-77-log-1283)": etcdserver: request timed out

script returned exit code 1
```

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.22

nixpanic commented:
CephFS NodeUnpublishVolume never returned ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/2274/pipeline)):
```
Mar 17 07:30:56.406: INFO: PersistentVolume pvc-80c82e0f-f9a8-4554-8c6b-db975cb13967 found and phase=Bound (2.645108ms)
Mar 17 07:30:56.422: INFO: Waiting up to csi-cephfs-demo-pod to be in Running state
Mar 17 07:31:02.438: INFO: Waiting for pod csi-cephfs-demo-pod to be deleted
Mar 17 07:41:02.452: FAIL: failed to validate CephFS pvc and application binding: timed out waiting for the condition
...
I0317 07:33:05.369623       1 utils.go:191] ID: 52 Req-ID: 0001-0024-2fe35e4b-d251-4e22-a116-bd6496352ee9-0000000000000001-2deb9525-a5c4-11ec-afd0-4e1bf5dd6186 GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0317 07:33:05.369732       1 utils.go:195] ID: 52 Req-ID: 0001-0024-2fe35e4b-d251-4e22-a116-bd6496352ee9-0000000000000001-2deb9525-a5c4-11ec-afd0-4e1bf5dd6186 GRPC request: {"target_path":"/var/lib/kubelet/pods/b205ec8b-88e3-4feb-8619-f68eafec58b3/volumes/kubernetes.io~csi/pvc-80c82e0f-f9a8-4554-8c6b-db975cb13967/mount","volume_id":"0001-0024-2fe35e4b-d251-4e22-a116-bd6496352ee9-0000000000000001-2deb9525-a5c4-11ec-afd0-4e1bf5dd6186"}
I0317 07:33:05.380060       1 mount_linux.go:218] Cannot run systemd-run, assuming non-systemd OS
I0317 07:33:05.380097       1 mount_linux.go:219] systemd-run output: System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to create bus connection: Host is down
, failed with: exit status 1
```

Rakshith-R commented:
@mergifyio refresh
